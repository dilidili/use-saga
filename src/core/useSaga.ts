import { useReducer, Reducer, useEffect, useRef } from 'react';
import { runSaga, stdChannel, Saga } from 'redux-saga';
import { takeEvery, call, put } from 'redux-saga/effects'

interface Action {
  type: string;
  [key: string]: any;
};

type Model<S> = {
  state: any,
  reducers?: {
    [key: string]: Reducer<S, Action>;
  },
  effects?: {
    [key: string]: Saga;
  },
}


export default function useSaga<S = {}>(model: Model<S>): [S, (action: Action) => void] {
  const {
    state: initialState,
    reducers = {},
    effects = {},
  } = model;

  // reducers
  const reducer = (prevState: S, action: Action) => Object.keys(reducers).reduce((r, v) => v === action.type ? reducers[v](prevState, action) : r, prevState);
  const [state, dispatch] = useReducer(reducer, initialState);

  // effects
  const refState = useRef<S>(state);
  const refDispatch = useRef(dispatch);
  const refChannel = useRef(stdChannel());

  useEffect(() => {
    refState.current = state; 
    refDispatch.current = dispatch; 
  }, [state, dispatch]);

  useEffect(() => {
    const boundRunSaga = runSaga.bind(null, {
      channel: refChannel.current,
      dispatch: (...args) => refDispatch.current.apply(null, args),
      getState: () => refState.current,
    });

    boundRunSaga(function *() {
      const keys = Object.keys(effects);
      for(let i = 0; i < keys.length; i++) {
        yield takeEvery(keys[i], function* (action) {
          yield effects[keys[i]](action, { call, put });
        });
      }
    });
  }, []);

  return [state, (action: Action) => {
    refChannel.current.put(action)
  }];
};