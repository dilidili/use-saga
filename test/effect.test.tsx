import useSaga, { useFetchPlugin, useFetchState } from '../src';
import React from 'react';
import renderer from 'react-test-renderer';

test('put effect can trigger another effect', () => {
  const Test: React.FC = () => {
    const [{ count, }, dispatch] = useSaga({
      state: {
        count: 0,
      },

      effects: {
        toOne: function *(_, { put, setState }) {
          yield put({
            type: 'toTwo',
          });
        },
        toTwo: function *(_, { setState }) {
          yield setState({ count: 2 });
        },
      },
    });

    return <div onClick={() => dispatch({ type: 'toOne', })}>
      <div>{count}</div>
    </div>
  };

  let component
  renderer.act(() => {
    component = renderer.create(<Test />);
  });
  let tree = component.toJSON();
  renderer.act(() => {
    tree.props.onClick();
  });
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('useFetch plugin', () => {
  const Test: React.FC = () => {
    const [fetchUserActions, fetchUserPlugin] = useFetchPlugin('fetchUser', function *(id) { return { id, } });

    const [state, dispatch] = useSaga<{
      fetchUser: useFetchState,
    }>({
      plugins: [
        fetchUserPlugin,
      ],
    });

    return <div onClick={() => dispatch(fetchUserActions.call('target'))}>
      <div>{state.fetchUser.data ? state.fetchUser.data.id : ''}</div>
    </div>
  };

  let component
  renderer.act(() => {
    component = renderer.create(<Test />);
  });
  let tree = component.toJSON();
  renderer.act(() => {
    tree.props.onClick();
  });
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});