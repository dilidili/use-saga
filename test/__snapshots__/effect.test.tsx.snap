// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`put effect can trigger another effect 1`] = `
<div
  onClick={[Function]}
>
  <div>
    2
  </div>
</div>
`;

exports[`useFetch plugin 1`] = `
<div
  onClick={[Function]}
>
  <div>
    target
  </div>
</div>
`;
